src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/IProjectHotReloadUpdateApplier.cs

```diff
@@ -0,0 +1,26 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using Microsoft.VisualStudio.HotReload.Components.DeltaApplier;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.HotReload
+{
+    /// <summary>
+    /// Provides a mechanism to apply custom updates to the running hot reload sessions. Assumes all updates can be applied via the 
+    /// IDeltaApplier created for this session
+    /// </summary>
+    [ProjectSystemContract(ProjectSystemContractScope.UnconfiguredProject, ProjectSystemContractProvider.System, Cardinality = Composition.ImportCardinality.ExactlyOne)]
+    public interface IProjectHotReloadUpdateApplier
+    {
+        /// <summary>
+        /// Returns true if there is at least one hot reload session active in the project
+        /// </summary>
+        bool HasActiveHotReloadSessions { get; }
+
+        /// <summary>
+        /// Called to apply a custom update via the IDeltaApplier. The applyFunction will be called once for every active session, and passed the IDeltaApplier
+        /// for that session. In the case of multiple sessions, updates will stop being applied on the first call to an applyFunction that throws an exception, and 
+        /// that exception propagates back to the caller.
+        /// </summary>
+        Task ApplyHotReloadUpdateAsync(Func<IDeltaApplier, CancellationToken, Task> applyFunction, CancellationToken cancelToken);
```

Comment:
Would the caller want to know if this actually applied a change? This could return `false` if no session exists, for example.

Reply:
We could have the function return an array of results, where the result type is user-defined. Something like:

`Task<T[]> ApplyHotReloadUpdateAsync<T>(Func<IDeltaApplier, CancellationToken, Task<T>> applyFunction, CancellationToken cancellationToken)`

The resulting array would be empty if there were no active sessions, and if there were issues applying the change you could get back some information about them. Would that be useful?

Reply:
See my previous comment. It is important to know ahead of time whether processing needs to occur

Reply:
It's fine as-is, so long as the caller doesn't need to know 100% whether the update was applied. There is a race condition here.